### PR TITLE
provider/aws: aws_main_route_table_association

### PR DESCRIPTION
**Pull request by phinze**
This resource allows an existing Route Table to be assigned as the
"main" Route Table of a VPC. This means that the Route Table will be
used for any subnets within the VPC without an explicit Route Table
assigned [1].

This is particularly useful in getting an Internet Gateway in place as
the default for a VPC, since the automatically created Main Route Table
does not have one [2].

Note that this resource is an abstraction over an association and does not
map directly to a CRUD-able object in AWS. In order to retain a coherent
"Delete" operation for this resource, we remember the ID of the AWS-created
Route Table and reset the VPC's main Route Table to it when this
resource is deleted.

refs #843, #748

[1] http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_Route_Tables.html#RouteTableDetails
[2] http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_Internet_Gateway.html#Add_IGW_Routing

### CODE REVIEW

**Comment by phinze** [comment]
Ok paging @mitchellh for review. Happy to walk you through it synchronously if that would be helpful.

Working on docs now.

File: builtin/providers/aws/resource_aws_main_route_table_association_test.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+// Need to keep the old route table around when we update the
+// main_route_table_association, otherwise Terraform will try to destroy the
+// route table too early, and will fail because it's still the main one
+resource "aws_route_table" "foo" {
+	vpc_id = "${aws_vpc.foo.id}"
+	route {
+		cidr_block = "10.0.0.0/8"
+		gateway_id = "${aws_internet_gateway.foo.id}"
+	}
+}
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_main_route_table_association_test.go]: @mitchellh ^^ this limitation of the current terraform graph is worth noting. AFAICT there's currently no way of expressing the required order of operations to make this possible

**Comment by mitchellh** [comment]
This is actually perfect. Great job :+1: MERGE.

File: builtin/providers/aws/resource_aws_main_route_table_association_test.go (hunk 1 above)
Comment (mitchellh) [inline builtin/providers/aws/resource_aws_main_route_table_association_test.go]: I think this relates back to that same issue of implicit deps...

Comment (phinze) [inline builtin/providers/aws/resource_aws_main_route_table_association_test.go]: Yep exactly. Will be an interesting problem to solve! :thought_balloon:

**Comment by ptrkdy** [comment]
I was stuck on topic this and found the solution in this PR before I found the documentation (which is my own fault admittedly), but I would like to leave a bread crumb for the next person who might be searching for this.

https://www.terraform.io/docs/providers/aws/r/main_route_table_assoc.html
together with
https://www.terraform.io/docs/providers/aws/r/route_table.html

Yields something along the lines of the below (as of version 0.9.2):

---

resource "aws_route_table" "main_dev" {
    vpc_id = "${aws_vpc.main.id}"
    route {
        cidr_block = "0.0.0.0/0"
        gateway_id = "${aws_internet_gateway.dev_igw.id}"
    }
    tags {
        Name = "main"
    }
}

resource "aws_main_route_table_association" "main_dev" {
   vpc_id = "${aws_vpc.main.id}"
   route_table_id = "${aws_route_table.main_dev.id}"
}

---

When the main route column in the AWS console updated to 'Yes' it gave me a smile!

**Pull request merged**